src/rom_main.cc

```diff
@@ -23,76 +23,123 @@
-#include "SparseDistMatrix.h"
-#include "magma_singleton.h"
-#include "tools.h"
-
-#include <fenv.h>
-#include <sys/cdefs.h>
-#include <time.h>
-
-#include <boost/program_options.hpp>
-namespace po = boost::program_options;
-
-#include "librom.h"
+#include "rom_workflows.h"
 
 //#include "MemTrack.h"
 
-int main(int argc, char** argv)
-{
-    // change handling of memory allocation errors
-    set_new_handler(noMoreMemory);
+/*
+void trapfpe () {
+  feenableexcept(FE_INVALID|FE_DIVBYZERO|FE_OVERFLOW);
+}
+*/
 
-    cout.sync_with_stdio();
+// A helper function
+template <class T>
+std::ostream& operator<<(std::ostream& os, const std::vector<T>& v)
+{
+    copy(v.begin(), v.end(), std::ostream_iterator<T>(std::cout, " "));
+    return os;
+}
 
+int main(int argc, char** argv)
+{
     int mpirc = MPI_Init(&argc, &argv);
     if (mpirc != MPI_SUCCESS)
     {
-        cerr << "MPI Initialization failed!!!" << endl;
+        std::cerr << "MPI Initialization failed!!!" << std::endl;
         MPI_Abort(MPI_COMM_WORLD, 0);
     }
-    MPI_Comm_rank(MPI_COMM_WORLD, &mype);
-    assert(mype > -1);
-    onpe0 = (mype == 0);
 
-    CAROM::Vector librom_vector(10, false);
+    MPI_Comm comm = MPI_COMM_WORLD;
+
+    mgmol_init(comm);
+
+    // read runtime parameters
+    std::string input_filename("");
+    std::string lrs_filename;
+    std::string constraints_filename("");
+    bool tcheck = false;
+
+    float total_spin = 0.;
+    bool with_spin   = false;
+
+    po::variables_map vm;
+
+    // use configure file if it can be found
+    // std::string config_filename("mgmol.cfg");
+
+    // read options from PE0 only
+    if (MPIdata::onpe0)
+    {
+        read_config(argc, argv, vm, input_filename, lrs_filename,
+            constraints_filename, total_spin, with_spin, tcheck);
+    }
+
+    MGmol_MPI::setup(comm, std::cout, with_spin);
+    MGmol_MPI& mmpi      = *(MGmol_MPI::instance());
+    MPI_Comm global_comm = mmpi.commGlobal();
+
+    Control::setup(global_comm, with_spin, total_spin);
+    Control& ct = *(Control::instance());
+
+    ct.setOptions(vm);
+
+    int ret = ct.checkOptions();
+    if (ret < 0) return ret;
+
+    mmpi.bcastGlobal(input_filename);
+    mmpi.bcastGlobal(lrs_filename);
+
+    // Enter main scope
+    {
+        MGmolInterface* mgmol;
+        if (ct.isLocMode())
+            mgmol = new MGmol<LocGridOrbitals>(global_comm, *MPIdata::sout);
+        else
+            mgmol
+                = new MGmol<ExtendedGridOrbitals>(global_comm, *MPIdata::sout);
+
+        unsigned ngpts[3]    = { ct.ngpts_[0], ct.ngpts_[1], ct.ngpts_[2] };
+        double origin[3]     = { ct.ox_, ct.oy_, ct.oz_ };
+        const double cell[3] = { ct.lx_, ct.ly_, ct.lz_ };
+        Mesh::setup(mmpi.commSpin(), ngpts, origin, cell, ct.lap_type);
+
+        mgmol->setupFromInput(input_filename);
+
+        if (ct.isLocMode() || ct.init_loc == 1) mgmol->setupLRs(lrs_filename);
+
+        mgmol->setupConstraintsFromInput(constraints_filename);
```

Comment:
This line is not needed (only for MD)

Reply:
Are we limiting ROM development only to MD, or do we consider extending it to other types of simulation? If we do consider extensions, wouldn't it be good to keep this the same as `src/main.cc`? We're parsing `constraints_filename` from `read_config` anyway.

Reply:
constraints referred to atomic positions. you are not doing anything with it later in that driver, so it remains unused. just trying to keep it clean

Reply:
I agree it is good to keep it simple. My understanding is that this driver reads orbitals data from the restart file and generates the POD basis data. It is reasonable to limit it to just what it needs to complete these tasks. The driver can certainly be extended  as we get additional mgmol/libROM integration capabilities completed. At that point, we can bring in the other calls that are needed to make the process more systematic and help drive how this driver evolves.

Reply:
Currently it does not do anything about atomic positions. But I'm planning to implement ROM MD simulation as one mode of this driver. I *think* `MGmol<OrbitalsType>::md` updates constraints or at least use the constraints, is it correct? If so, the ROM version `md` will also need the constraints as well.

Reply:
Oh, I see